ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/ChainState.hs

```diff
@@ -490,3 +548,88 @@ instance Serialise (PBftVerKeyHash c) => Serialise (PBftSigner c) where
   decode = fromPair <$> decode
     where
       fromPair (slotNo, genesisKey) = PBftSigner slotNo genesisKey
+
+{-------------------------------------------------------------------------------
+  EBB Map
+-------------------------------------------------------------------------------}
+
+-- | Append an EBB
+--
+-- Its slot will be mapped to the 'lastSignedSlot'.
+appendEBB :: forall c. (PBftCrypto c, HasCallStack)
+          => SecurityParam
+          -> WindowSize
+          -> SlotNo
+          -> PBftChainState c -> PBftChainState c
+appendEBB k n newEbbSlot cs@PBftChainState{..} =
+    assertInvariant k n $
+    Exn.assert valid $
+    cs{ebbs = ebbsInsert newEbbSlot latestNonEbbSlot ebbs}
+  where
+    latestEbbSlot    = ebbsMax ebbs
+    latestNonEbbSlot = lastSignedSlot cs
+
+    valid = At newEbbSlot > max latestEbbSlot latestNonEbbSlot
+
+-- | Discard 'ebbs' mappings whose /value/ is before the anchor
+--
+-- Called by 'append', since 'ebbs' do not increase how far back the chain
+-- state can rewind.
+pruneEBBsLT :: PBftChainState c -> PBftChainState c
+pruneEBBsLT cs@PBftChainState{..} =
+  cs{ebbs = ebbsPruneValueLT (anchorSlot cs) ebbs}
+
+-- | Discard 'ebbs' mappings whose /key/ is after the given slot
+--
+-- Called by 'rewind', since 'rewind'ing to a slot should forget the EBBs it
+-- precedes.
+pruneEBBsGT :: WithOrigin SlotNo -> PBftChainState c -> PBftChainState c
+pruneEBBsGT mSlot cs@PBftChainState{..} =
+  cs{ebbs = ebbsPruneKeyGT mSlot ebbs}
+
+-- | A map from the slots containing an EBB to the preceding signed slot
+--
+-- INVARIANT Key greater: For all @(k, v)@, @At k > v@.
+--
+-- INVARIANT Non-descending: For all @(k1, v1)@ and @(k2, v2)@, @k1 < k2@
+-- implies @v1 <= v2@.
+newtype EbbMap = EbbMap {unEbbMap :: Map SlotNo (WithOrigin SlotNo)}
+  deriving stock (Generic, Show)
+  deriving newtype (Eq, Ord, NoUnexpectedThunks, Serialise)
+
+ebbsEmpty :: EbbMap
+ebbsEmpty = EbbMap Map.empty
+
+ebbsInsert :: SlotNo -> WithOrigin SlotNo -> EbbMap -> EbbMap
+ebbsInsert k v = EbbMap . Map.insert k v . unEbbMap
+
+ebbsMax :: EbbMap -> WithOrigin SlotNo
+ebbsMax = maybe Origin (At . fst) . Map.lookupMax . unEbbMap
+
+ebbsLookup :: SlotNo -> EbbMap -> Maybe (WithOrigin SlotNo)
+ebbsLookup k = Map.lookup k . unEbbMap
+
+-- | Equivalent to @\mSlot -> Map.filter (>= mSlot)@
+ebbsPruneValueLT :: WithOrigin SlotNo -> EbbMap -> EbbMap
+ebbsPruneValueLT mSlot0 (EbbMap m0) =
+    EbbMap $
+    case mSlot0 of
+      Origin -> m0   -- @_ >= Origin@ always holds
+      At s   -> go gt
+        where
+          -- NB the \"Key greater\" invariant
+          gt = snd $ Map.split s m0
+  where
+    -- NB the \"Non-descending\" invariant
+    go m = case Map.minViewWithKey m of
+      Just ((_, mSlot), m')
+        | mSlot < mSlot0 -> go m'
+      _                  -> m
```

Comment:
Can you add a property test checking that this function is equivalent to `\mSlot -> Map.filter (>= mSlot)`?

Since the `EbbMap` will in practice contain at most one entry (one EBB every `10k` blocks and the window size is `k`, I believe), I think we should actually just use `Map.filter` as it is much easier to read, evidently correct, and probably not slower than this approach.

Reply:
I added the tests for both. Yes, I was worried the functions were overkill, at least for production. I suppose they might accelerate some hypothetical tests with small epochs?

Reply:
Fixed, thanks.